Mathlib/RingTheory/FinitePresentation.lean

```diff
@@ -450,6 +450,57 @@ end FinitePresentation
 
 end RingHom
 
+namespace RingHom.FinitePresentation
+universe u v
+variable {R : Type u} [CommRing R]
+
+open Polynomial
+
+/-- Induction principle for finitely presented ring homomorphisms.
+
+For a property to hold for all finitely presented ring homs, it suffices for it to hold for
+`Polynomial.C : R → R[X]`, surjective ring homs with finitely generated kernels, and to be closed
+under composition. -/
+lemma polynomial_induction
+    (P : ∀ (R : Type u) [CommRing R] (S : Type u) [CommRing S], (R →+* S) → Prop)
+    (Q : ∀ (R : Type u) [CommRing R] (S : Type v) [CommRing S], (R →+* S) → Prop)
+    (polynomial : ∀ (R) [CommRing R], P R R[X] C)
+    (fg_ker : ∀ (R : Type u) [CommRing R] (S : Type v) [CommRing S] (f : R →+* S),
+      Surjective f → (ker f).FG → Q R S f)
+    (comp : ∀ (R) [CommRing R] (S) [CommRing S] (T) [CommRing T] (f : R →+* S) (g : S →+* T),
+      P R S f → Q S T g → Q R T (g.comp f))
+    {R S} [CommRing R] [CommRing S] (f : R →+* S) (hf : f.FinitePresentation) :
+    Q R S f := by
+  obtain ⟨n, g, hg, hg'⟩ := hf
+  let g' := letI := f.toAlgebra; g.toRingHom
+  replace hg : Surjective g' := hg
+  replace hg' : (ker g').FG := hg'
+  have : g'.comp MvPolynomial.C = f := letI := f.toAlgebra; g.comp_algebraMap
```

Comment:
seems cleaner to have `letI := f.toAlgebra` outside, since you use it twice.
Or does that break for some reason?